SCMI server for OP-TEE, Zephyr and Vhost User

vingu-linaro opened the pull request:
This P-R integrates the changes that we have made in SCP firmware in order to make it run as a SCMI server in:

-  OP-TEE env. This has been tested on FVP arm64 and stm32. SCP-Firmware is integrated as a sub-module of op_tee-os. this PR has been tested with : https://github.com/vingu-linaro/optee_os/tree/scmi-over-optee. The product/optee-fvp is used with fvp platform.

- Vhost user virtio SCMI backend. The SCP-firmware handles virtio-scmi request over a vhost user socket. This has been tested with a Linux image VM with Qemu. The product/vhost-user is built to monitor vscmi.ospm0.sock and vscmi.ospm1.sock. Additional Qemu parameter to provides: 
qemu-system-aarch64  <usual qemu parameters to run rou linux image ..> -chardev socket,path=vscmi-ospm0.sock,id=vscmi   -device vhost-user-scmi-device,chardev=vscmi,id=scmi  -object memory-backend-file,id=mem,size=1G,mem-path=/dev/shm,share=on -numa node,memdev=mem -dtb /home/vingu/Linaro/Kernel/out/binaries/arm64/virt-scmi.dtb -global virtio-mmio.force-legacy=false. You have to create your own dtb file based on Qemu's one in order to add the scmi protocol that you want to enable in Linux.

- Virtio MMIO/ virtio SCMI device in zephyr. The SCP-firmware is used as a submodule of a zephyr app : https://github.com/vingu-linaro/mySCP-zephyr. It handles virtio-mmio request through a new virtio mmio zephyr device. This full setup is a bit  more complex to put in place as it needs some changes in several other repo but i can provides the details if you are interested in. This one uses CMake

I don't expect this P-R to be merged as it is but appreciate some feedback about the way the new arch have been added in arch/none. The P-R also add 2 new features:
- add a framework stop stage to let module and element to cleanly destroy and release resource and memory. This is useful for the vhost user backend as it delete cleanly host resources
- add a context in mono threading mode. Enable to create an event list per channel. The goal is to be able to handle simultaneously and independently the requests from 2 differents channels which don't access the same resources.

girishpathak commented:
@vingu-linaro: We are reviewing the patchset, However, we have noticed few changes which can be done without too many changes to the existing arch code. I've created a patch on top of your patchset, it basically reverts some changes and adds bit of new changes. I've not compiled it as is intended only for discussion. 
[0001-REVIEW-scmi-over-optee-Arch-changes-suggestions.txt](https://github.com/ARM-software/SCP-firmware/files/7287854/0001-REVIEW-scmi-over-optee-Arch-changes-suggestions.txt)

girishpathak commented:
@vingu-linaro: can you provide a bit more details on Zephyr and SCP firmware integration ? any document ? repositories ?

vingu-linaro commented:
> @vingu-linaro: We are reviewing the patchset, However, we have noticed few changes which can be done without too many changes to the existing arch code. I've created a patch on top of your patchset, it basically reverts some changes and adds bit of new changes. I've not compiled it as is intended only for discussion. [0001-REVIEW-scmi-over-optee-Arch-changes-suggestions.txt](https://github.com/ARM-software/SCP-firmware/files/7287854/0001-REVIEW-scmi-over-optee-Arch-changes-suggestions.txt)
> 
> Thoughts ?

Hi Girish,

Thanks for the review. I haven't look in details your changes but looks reasonable to me. I have some pending changes that i need to push for Zephyr. I will do by end of this week

vingu-linaro commented:
> @vingu-linaro: can you provide a bit more details on Zephyr and SCP firmware integration ? any document ? repositories ?

I will add details about how to build and run scp-firmware with zephyr in a VM while pushing the update end of this week

vingu-linaro commented:
Hi,

I have pushed a new version of the PR. This PR doesn't include girishpathak's suggestion yet.

In order to run SCMI-backend based on zephyr in a guest VM, you need to follow the zephyr how-to in order to install zephyrproject , zephyr-sdk ... Then you have to download:

**MySCP-zephyr application:
Simple zephyr application that integrates SCP-firmware as a submodule in a zephyr application dev environnement

url: https://github.com/vingu-linaro/mySCP-zephyr
branch scmi-over-zephyr

**SCP-firmware (scmi server)
This repo will be downloaded as a submodule of mySCP-zephyr
https://github.com/vingu-linaro/SCP-firmware.git
branch: scmi-over-zephyr

In zephyrproject,  you have to use the following branches:

**Zephyr:
Add support virtio-mmio and virtio-scmi device in zephyr
https://github.com/vingu-linaro/zephyr
branch scmi-over-zephyr

**Open-amp:
https://github.com/vingu-linaro/open-amp
branch scmi-over-zephyr

** libmetal
https://github.com/vingu-linaro/libmetal
branch scmi-over-zephyr

You also need a Linux that includes virtio-scmi FE
**Linux:
Support of virtio-scmi is part of v515-rc4

And you need the vhost user bridge:
**Qemu:
vhost user scmi device + bridge
https://git.linaro.org/people/vincent.guittot/qemu.git
branch: vhost-user-scmi

**Instructions:
-download and compile Qemu to get the vhost-user-scmi bridge
#vhost-user-scmi --socket-path-vm vscmi-guest.sock --socket-path-be
<path to MySCP-zephyr application>/vscmi-server.sock

-download and compile Linux with virtio scmi FE
use vscmi-guest.sock in the qemu parameter when starting your linux guest VM

- download and compile MySCP-zephyr application:
Once downloaded, use west tool (zephyr wrapper over cmake)
#west build -t run -b qemu_cortex_a53  -- -G"Unix Makefiles"
-DSCP_FIRMWARE_SOURCE_DIR:PATH=<path to MySCP-zephyr
application>/src/SCP-firmware/product/virtio-zephyr/fw/
-DSCP_LOG_LEVEL=INFO -DCMAKE_OBJCOPY:FILEPATH=<path to zephyr
sdk>/aarch64-zephyr-elf/bin/aarch64-zephyr-elf-objcopy
-DCMAKE_OBJDUMP:FILEPATH=<path to zephyr
sdk>/aarch64-zephyr-elf/bin/aarch64-zephyr-elf-objdump

This will compile the scmi-backend and then call qemu to launch the scmi-zephyr VM. By default Qemu looks for a socket in <path to MySCP-zephyr application>/

girishpathak commented on module/clock/src/mod_clock.c:
```diff
@@ -366,9 +369,11 @@ static int clock_init(fwk_id_t module_id, unsigned int element_count,
         return FWK_SUCCESS;
     }
 
+#ifdef BUILD_HAS_NOTIFICATION
```
is this change required ?

vingu-linaro replied:
struct mod_clock_config {
    const fwk_id_t pd_transition_notification_id;
    const fwk_id_t pd_pre_transition_notification_id;
};

Clock config is only used to set notification id and is null when notification is not supported

girishpathak commented on module/clock/src/mod_clock.c:
```diff
@@ -30,7 +30,7 @@ static struct clock_ctx module_ctx;
  * Utility functions
  */
-
+#ifdef BUILD_HAS_NOTIFICATION
```
Is this change required ?, I think process_response_event don't  come in notifications path

vingu-linaro replied:
MOD_CLOCK_API_TYPE_DRIVER_RESPONSE --> clock_driver_response_api --> clock_request_complete --> mod_clock_event_id_response --> CLOCK_EVENT_IDX_RESPONSE --> process_response_event

This is the only use that I have found and  MOD_CLOCK_API_TYPE_DRIVER_RESPONSE is under #if defined(BUILD_HAS_NOTIFICATION). 
I have compilation error for unused variable / function

tarek-arm replied:
Yes, but why did you put MOD_CLOCK_API_TYPE_DRIVER_RESPONSE is under #if defined(BUILD_HAS_NOTIFICATION)?

vingu-linaro replied:
so I re run some test without notification and the line above and the one for MOD_CLOCK_API_TYPE_DRIVER_RESPONSE is not necessary.
Will remove it

girishpathak commented on module/power_domain/src/mod_power_domain.c:
```diff
@@ -754,8 +764,12 @@ static void respond(struct pd_ctx *pd, int resp_status)
         return;
     }
 
+#ifdef BUILD_HAS_NOTIFICATION
```
this is not required. fwk_thread_get_delayed_response is separate than notifications

vingu-linaro replied:
yes you're right

girishpathak commented on module/power_domain/src/mod_power_domain.c:
```diff
@@ -930,10 +944,15 @@ static void process_set_state_request(
     }
 
     if (pd_in_charge_of_response != NULL) {
+#ifdef BUILD_HAS_NOTIFICATION
```
this is not required. fwk_thread_get_delayed_response is separate than notifications

vingu-linaro replied:
yes

girishpathak commented on module/power_domain/src/mod_power_domain.c:
```diff
@@ -1195,6 +1218,7 @@ static void process_power_state_transition_report(struct pd_ctx *pd,
         return;
     }
 
+#ifdef BUILD_HAS_NOTIFICATION
```
I think, Below code is not related to notifications

if (pd->parent == NULL) {
/* this is the top pd (SYSTOP) /
if (mod_pd_ctx.system_suspend.state != MOD_PD_STATE_ON) {
/ has gone down, invalidate the system suspend ongoing */
mod_pd_ctx.system_suspend.suspend_ongoing = false;
}

vingu-linaro replied:
yes you're right

girishpathak commented on module/power_domain/src/mod_power_domain.c:
```diff
@@ -1351,9 +1376,11 @@ void perform_shutdown(
      * be notified.
      */
     if (resp == NULL) {
+#ifdef BUILD_HAS_MULTITHREADING
```
unreleated to multi-threading

girishpathak commented on framework/src/fwk_multi_thread.c:
```diff
@@ -252,6 +252,7 @@ static int put_event(
     }
 
     if (std_event != NULL && std_event->is_delayed_response) {
+#ifdef BUILD_HAS_NOTIFICATION
```
unrelated to notifications

vingu-linaro replied:
This is an old commit that has been done when starting to work op-tee architecture without notification. But this doesn't seem necessary anymore. Will remove it

girishpathak commented on arch/arm/armv7-m/src/arch_main.c:
```diff
@@ -54,7 +54,19 @@ static void arch_init_ccr(void)
     arch_init_ccr();
 
-    return fwk_arch_init(&arch_init_driver);
+    status = fwk_arch_init(&arch_init_driver);
```
see https://github.com/ARM-software/SCP-firmware/pull/524#issuecomment-934602620)

vingu-linaro replied:
Yes, I'm going to look more closely on your proposal and integrate it in the next version

vingu-linaro replied:
I have  studied your proposal and integrated it in the following way:
- armv7-m and armv-a are not modified by this PR anymore

- I have renamed SCP_ARCHITECTURE_VM into SCP_ARCHITECTURE_OS and BUILD_HAS_ARCH_VM into BUILD_HAS_ARCH_OS because zephyr and optee are not VM (but our main use case with zephyr is to run it into a VM)

- I have kept __fwk_thread_run calling __fwk_run_event in order to avoid duplicating the same code twice. The end result is the same for __fwk_thread_run() which is : 
  1- loop on ctx.event_queue and process event until the list is empty
  2- process_isr and continue to step 1 if there is an isr event
  3- unbuffer log 
  4- restart on 1

I have run few more test and will push the new version by end of the day

girishpathak commented on arch/arm/armv8-a/src/arch_main.c:
```diff
@@ -48,4 +48,9 @@ void arm_main(void)
     status = fwk_arch_init(&arch_init_driver);
     if (status != FWK_SUCCESS)
         panic();
+
```
see https://github.com/ARM-software/SCP-firmware/pull/524#issuecomment-934602620

girishpathak reviewed (commented):
Please see some of the comments below

vingu-linaro commented on module/scmi_perf/src/mod_scmi_perf.c:
```diff
@@ -1924,6 +1926,7 @@ static int scmi_perf_start(fwk_id_t id)
      */
     if (!fwk_id_is_equal(scmi_perf_ctx.config->fast_channels_alarm_id,
         FWK_ID_NONE)) {
+#if BUILD_HAS_MOD_TIMER
```
yes, i will include the if condition

vingu-linaro commented on module/dvfs/src/mod_dvfs.c:
```diff
@@ -652,10 +668,12 @@ static int dvfs_complete(
     }
 
     /* notify the HAL that the level has been updated */
+#ifdef BUILD_HAS_NOTIFICATION
```
Not sure to get the comment about single-threaded

Nevertheless, #ifdef BUILD_HAS_NOTIFICATION is not needed

vingu-linaro commented on module/dvfs/src/mod_dvfs.c:
```diff
@@ -582,8 +592,10 @@ static void dvfs_complete_respond(
     struct fwk_event *resp_event,
     int req_status)
 {
+#ifdef BUILD_HAS_MULTITHREADING
```
yes. will fix it

vingu-linaro commented:
I have rebased and tested on latest version of master branch and tried to address all comments made on previous version

girishpathak commented:
@vingu-linaro : I am getting error when I execute below

#west build -t run -b qemu_cortex_a53 -- -G"Unix Makefiles"
-DSCP_FIRMWARE_SOURCE_DIR:PATH=/src/SCP-firmware/product/virtio-zephyr/fw/
-DSCP_LOG_LEVEL=INFO -DCMAKE_OBJCOPY:FILEPATH=/aarch64-zephyr-elf/bin/aarch64-zephyr-elf-objcopy
-DCMAKE_OBJDUMP:FILEPATH=/aarch64-zephyr-elf/bin/aarch64-zephyr-elf-objdump

devicetree error: 'reg' property in <Node /virtio_mmio@a003e00 in 'qemu_cortex_a53.dts.pre.tmp'> has length 8, which is not evenly divisible by 16 (= 4*(<#address-cells> (= 2) + <#size-cells> (= 2))). Note that #*-cells properties come either from the parent node or from the controller (in the case of 'interrupts').

vingu-linaro commented:
> @vingu-linaro : I am getting error when I execute below
> 
> #west build -t run -b qemu_cortex_a53 -- -G"Unix Makefiles" -DSCP_FIRMWARE_SOURCE_DIR:PATH=/src/SCP-firmware/product/virtio-zephyr/fw/ -DSCP_LOG_LEVEL=INFO -DCMAKE_OBJCOPY:FILEPATH=/aarch64-zephyr-elf/bin/aarch64-zephyr-elf-objcopy -DCMAKE_OBJDUMP:FILEPATH=/aarch64-zephyr-elf/bin/aarch64-zephyr-elf-objdump
> 
> devicetree error: 'reg' property in <Node /virtio_mmio@a003e00 in 'qemu_cortex_a53.dts.pre.tmp'> has length 8, which is not evenly divisible by 16 (= 4*(<#address-cells> (= 2) + <#size-cells> (= 2))). Note that #*-cells properties come either from the parent node or from the controller (in the case of 'interrupts').
> 
> Any idea ?

which dtc tool version are you using ?

> 
> Also can you please let me know where to checkout and how to build
> 
> ** open-amp https://github.com/vingu-linaro/open-amp branch scmi-over-zephyr
> 
> ** libmetal https://github.com/vingu-linaro/libmetal branch scmi-over-zephyr

When you have toinit and sync the zephyr project dir: named zephyrproject dir in zephyr how to, you will find the zephyr/ dir which contains the zephyr RTOS and also a modules/ directory where  you should find modules/hal/libmetal/ and modules/lib/open-amp/

girishpathak commented:
@vingu-linaro:  Hi,

> **Instructions:
> -download and compile Qemu to get the vhost-user-scmi bridge
> #vhost-user-scmi --socket-path-vm vscmi-guest.sock --socket-path-be
> /vscmi-server.sock

> -download and compile Linux with virtio scmi FE
> use vscmi-guest.sock in the qemu parameter when starting your linux guest VM

I was able to get vhost-user-scmi bridge build and running. I was also able to build the Linux kernel.

However following part 

> -download and compile Linux with virtio scmi FE
> use vscmi-guest.sock in the qemu parameter when starting your linux guest VM

Is not very clear to me.  Can you please elaborate on the steps to start Linux guest VM using qemu ?

girishpathak commented:
@vingu-linaro : I'm not sure if you noticed above  last week, just like to know  above steps for me to proceed.

vingu-linaro commented:
> @vingu-linaro : I'm not sure if you noticed above last week, just like to know above steps for me to proceed.

Sorry, your comment got lot in my inbox

Once you have compiled Linux with virtio scmi which is now supported mainline, you have to add a virtio-scmi when starting qemu. You have to use the same socket "vscmi-guest.sock" that is used by the bridge

A typical command:
qemu-system-aarch64	-kernel <path to linux image>/Image \
			-hda <path to your distro rootfs>/qemu-image-aarch64-1.img \
			-append 'root=/dev/vda console=ttyAMA0 loglevel=8 sched_debug nokaslr' -nographic \
			-machine virt,gic-version=max -m 1G -cpu cortex-a57 -smp 2 -semihosting-config enable=on,target=native \
			-numa node,memdev=mem \
			-device vhost-user-scmi-device,chardev=vscmi,id=scmi \
			-chardev socket,path=vscmi-guest.sock,id=vscmi \
			-object memory-backend-file,id=mem,size=1G,mem-path=/dev/shm,share=on \
			-global virtio-mmio.force-legacy=false \
			-dtb virt-scmi.dtb

you should also notice -dtb virt-scmi.dtb parameter. If you use the default dtb provied by qemu, virt-scmi device will be discovered but no scmi agent will be created on top. You must add a firmware node with scmi like:
	firmware {
		scmi0: scmi0 {
			compatible = "arm,scmi-virtio";
			#address-cells = <1>;
			#size-cells = <0>;

			scmi_devpd: protocol@11 {
				reg = <0x11>;
				#power-domain-cells = <1>;
			};

			scmi_clk0: protocol@14 {
				reg = <0x14>;
				#clock-cells = <1>;
			};

			scmi_sensors0: protocol@15 {
				reg = <0x15>;
				#thermal-sensor-cells = <1>;

			};

			scmi_dvfs: protocol@13 {
				reg = <0x13>;
				#clock-cells = <1>;
			};
		};
	};

I advice you to dump qemu dtb, add the firmware node and provide this new dtb using -dtb <your dtb>

vingu-linaro commented:
I have just push a new version of the branch. The main difference compared to previous one is located in the module for zephyr virtio-mmio transport channel.

girishpathak commented on arch/none/host/src/arch_interrupt.c:
```diff
@@ -14,77 +14,82 @@
 #include <stdbool.h>
 #include <stdint.h>
 
+/*
```
Do not change these files in arch/none/host, Can you add a new arch e.g. arch/none/host-scmi-server ?

girishpathak commented on arch/none/host/src/arch_main.c:
```diff
@@ -5,33 +5,111 @@
  * SPDX-License-Identifier: BSD-3-Clause
  */
 
+#include <glib.h>
```
Do not change these files in arch/none/host, Can you add a new arch e.g. arch/none/host-scmi-server ?

girishpathak commented on module/input_mhu/include/mod_input_mhu.h:
```diff
@@ -0,0 +1,25 @@
+/*
+ * Arm SCP/MCP Software
+ * Copyright (c) 2015-2019, Arm Limited and Contributors. All rights reserved.
```
2022

girishpathak commented on module/input_mhu/src/Makefile:
```diff
@@ -0,0 +1,11 @@
+#
+# Arm SCP/MCP Software
+# Copyright (c) 2017-2020, Arm Limited and Contributors. All rights reserved.
```
2022

girishpathak commented on module/input_mhu/src/mod_input_mhu.c:
```diff
@@ -0,0 +1,443 @@
+/*
+ * Arm SCP/MCP Software
+ * Copyright (c) 2015-2019, Arm Limited and Contributors. All rights reserved.
```
2022

girishpathak commented on module/input_mhu/src/mod_input_mhu.c:
```diff
@@ -0,0 +1,443 @@
+ * Input host helpers
+ */
+
+static int kbd_simple_msg(int proto, int msg, int token, struct mod_virtio_smt_memory *memory)
```
please do not allow lines more than 80 char here and everywhere below

girishpathak commented on module/input_mhu/src/mod_input_mhu.c:
```diff
@@ -0,0 +1,443 @@
+    } else    if (!strncmp(str,"sensor trip", 11 )) {
+        sscanf(str, "sensor trip %d %u", &i, &level);
+        return kbd_sensor_trip(0x15, 0x5, 0xFF, i, level, memory);
+    }
```
please add ending else statement 
else {
/* empty *
}
to satisfy misra-c

girishpathak commented on module/input_mhu/src/mod_input_mhu.c:
```diff
@@ -0,0 +1,443 @@
+        return kbd_sensor_trip(0x15, 0x5, 0xFF, i, level, memory);
+    }
+
+    return -1;
```
no magic number please

girishpathak commented on module/input_mhu/src/mod_input_mhu.c:
```diff
@@ -0,0 +1,443 @@
+    GError *error = NULL;
+    struct virt_msg msg;
+
+    if( g_io_channel_read_line( channel, &str_return, &length, &terminator_pos, &error) == G_IO_STATUS_ERROR )
```
use {}

girishpathak commented on module/input_mhu/src/mod_input_mhu.c:
```diff
@@ -0,0 +1,443 @@
+    msg.out_len = kbd_set_message(str_return, (struct mod_virtio_smt_memory *) &msg.out_buf);
+    msg.in_len = 128;
+
+    if (msg.out_len > 0)
```
use {}

girishpathak commented on module/input_mhu/src/mod_input_mhu.c:
```diff
@@ -0,0 +1,443 @@
+                      const void *data)
+{
+
+    if (device_count == 0)
```
use {}

girishpathak commented on module/input_mhu/src/mod_input_mhu.c:
```diff
@@ -0,0 +1,443 @@
+
+    mhu_ctx.device_ctx_table = fwk_mm_calloc(device_count,
+                                             sizeof(*mhu_ctx.device_ctx_table));
+    if (mhu_ctx.device_ctx_table == NULL)
```
use {}

girishpathak commented on module/input_mhu/src/mod_input_mhu.c:
```diff
@@ -0,0 +1,443 @@
+        for (slot = 0; slot < device_ctx->slot_count; slot++) {
+            smt_channel = &device_ctx->smt_channel_table[slot];
+
+            if (smt_channel->id.value == 0)
```
use {}

girishpathak commented on module/input_mhu/src/mod_input_mhu.c:
```diff
@@ -0,0 +1,443 @@
+                device_ctx->config->driver_api_id,
+                &smt_channel->api);
+
+            if (status != FWK_SUCCESS)
```
use {}

girishpathak commented on module/input_mhu/src/mod_input_mhu.c:
```diff
@@ -0,0 +1,443 @@
+}
+
+static int vhost_process_bind_request(fwk_id_t source_id,
+                      fwk_id_t target_id,
```
alignment seems off

girishpathak commented on module/input_mhu/src/mod_input_mhu.c:
```diff
@@ -0,0 +1,443 @@
+    struct mhu_device_ctx *device_ctx;
+    struct mhu_smt_channel *smt_channel;
+
+    if (!fwk_id_is_type(target_id, FWK_ID_TYPE_SUB_ELEMENT))
```
use {}

girishpathak commented on module/input_mhu/src/mod_input_mhu.c:
```diff
@@ -0,0 +1,443 @@
+    if (!fwk_id_is_type(target_id, FWK_ID_TYPE_SUB_ELEMENT))
+        return FWK_E_ACCESS;
+
+    if (fwk_id_get_api_idx(api_id) != 0)
```
use {} here and everywhere below

girishpathak commented on module/host_console/src/Makefile:
```diff
@@ -0,0 +1,11 @@
+#
+# Arm SCP/MCP Software
+# Copyright (c) 2017-2021, Arm Limited and Contributors. All rights reserved.
```
2022

girishpathak commented on module/host_console/src/mod_host_console.c:
```diff
@@ -0,0 +1,86 @@
+/*
+ * Arm SCP/MCP Software
+ * Copyright (c) 2017-2021, Arm Limited and Contributors. All rights reserved.
```
2022

girishpathak commented on product/host-input/fw/clock_devices.h:
```diff
@@ -0,0 +1,25 @@
+/*
+ * Arm SCP/MCP Software
+ * Copyright (c) 2017-2019, Arm Limited and Contributors. All rights reserved.
```
2022

girishpathak commented on product/host-input/fw/config_clock.c:
```diff
@@ -0,0 +1,102 @@
+/*
+ * Arm SCP/MCP Software
+ * Copyright (c) 2015-2019, Arm Limited and Contributors. All rights reserved.
```
2022

girishpathak commented on product/host-input/fw/config_dvfs.c:
```diff
@@ -0,0 +1,156 @@
+/*
+ * Arm SCP/MCP Software
+ * Copyright (c) 2017-2019, Arm Limited and Contributors. All rights reserved.
```
2022 here and everywhere below

girishpathak commented on product/host-input/fw/config_dvfs.h:
```diff
@@ -0,0 +1,18 @@
+/*
+ * Arm SCP/MCP Software
+ * Copyright (c) 2017-2019, Arm Limited and Contributors. All rights reserved.
```
2022

girishpathak commented on product/host-input/fw/config_log.c:
```diff
@@ -0,0 +1,10 @@
+/*
+ * Arm SCP/MCP Software
+ * Copyright (c) 2017-2019, Arm Limited and Contributors. All rights reserved.
```
2022

girishpathak commented on product/host-input/fw/config_mhu_smt.c:
```diff
@@ -0,0 +1,113 @@
+/*
+ * Arm SCP/MCP Software
+ * Copyright (c) 2019, Arm Limited and Contributors. All rights reserved.
```
2022

girishpathak commented on product/host-input/fw/config_mock_psu.c:
```diff
@@ -0,0 +1,58 @@
+/*
+ * Arm SCP/MCP Software
+ * Copyright (c) 2017-2019, Arm Limited and Contributors. All rights reserved.
```
2022

girishpathak commented on product/host-input/fw/config_mock_psu.h:
```diff
@@ -0,0 +1,20 @@
+/*
+ * Arm SCP/MCP Software
+ * Copyright (c) 2019-2020, Arm Limited and Contributors. All rights reserved.
```
2022

girishpathak commented on product/host-input/fw/config_power_domain.c:
```diff
@@ -0,0 +1,175 @@
+/*
+ * Arm SCP/MCP Software
+ * Copyright (c) 2017-2019, Arm Limited and Contributors. All rights reserved.
```
2022

girishpathak commented on product/host-input/fw/config_power_domain.c:
```diff
@@ -0,0 +1,175 @@
+#include "config_power_domain.h"
+#include "config_ppu_v0.h"
+
+/* Mask of the allowed states for the systop power domain */
```
remove commented code or else provide justification in a comment here and everywhere below

girishpathak commented on product/host-input/fw/config_power_domain.h:
```diff
@@ -0,0 +1,20 @@
+/*
+ * Arm SCP/MCP Software
+ * Copyright (c) 2018-2019, Arm Limited and Contributors. All rights reserved.
```
2022

girishpathak commented on arch/none/host/src/arch_main.c:
```diff
@@ -38,9 +38,7 @@ int scmi_arch_init(void)
-//  TBD: no way to close cleanly the scp firmware
-
-    return 0;
+    return fwk_arch_deinit();
```
again,  please create arch/none/host-scmi-server and implement/modify code there

girishpathak commented on framework/include/fwk_module.h:
```diff
@@ -221,6 +221,26 @@ struct fwk_module {
      */
     int (*start)(fwk_id_t id);
+    /*!
+     * \brief Pointer to the start function.
```
Pointer to the sop function.

girishpathak commented on framework/src/fwk_arch.c:
```diff
@@ -98,6 +98,17 @@ int fwk_arch_init(const struct fwk_arch_init_driver *driver)
+    int status;
+
+    status = fwk_module_stop();
+    if (!fwk_expect(status == FWK_SUCCESS))
```
use {}

girishpathak commented on framework/src/fwk_module.c:
```diff
@@ -451,6 +452,86 @@ int fwk_module_start(void)
+
+        if (module->stop != NULL) {
+            status = module->stop(
+					fwk_id_build_element_id(module_ctx->id, element_idx));
```
alignment seems off

girishpathak commented on framework/src/fwk_module.c:
```diff
@@ -451,6 +452,86 @@ int fwk_module_start(void)
+    for (module_idx = 0; module_idx < FWK_MODULE_IDX_COUNT; module_idx++) {
+        module_ctx = &fwk_module_ctx.module_ctx_table[module_idx];
+        status = fwk_module_stop_module(module_ctx);
+        if (status != FWK_SUCCESS)
```
use {}

girishpathak commented on module/virtio_smt/src/mod_virtio_smt.c:
```diff
@@ -0,0 +1,475 @@
+    channel_ctx =
+        &smt_ctx.channel_ctx_table[fwk_id_get_element_idx(channel_id)];
+
+    if (!channel_ctx->locked)
```
use {}

girishpathak commented on module/virtio_smt/src/mod_virtio_smt.c:
```diff
@@ -0,0 +1,475 @@
+    channel_ctx =
+        &smt_ctx.channel_ctx_table[fwk_id_get_element_idx(channel_id)];
+
+    if (!channel_ctx->locked)
```
use {}

girishpathak commented on module/vhost_mhu/src/mod_vhost_mhu.c:
```diff
@@ -0,0 +1,463 @@
+    }
+}
+
+void vhost_signal_message(struct mhu_device_ctx *device_ctx, unsigned int channel_idx, VuVirtqElement *elem)
```
seems more than 80 char

girishpathak commented on module/vhost_mhu/src/mod_vhost_mhu.c:
```diff
@@ -0,0 +1,463 @@
+
+static void vscmi_handle_request_ctrl(VuDev *dev, int qidx)
+{
+    struct mhu_device_ctx *device_ctx = container_of(dev, struct mhu_device_ctx, dev.parent);
```
seems more than 80 char

girishpathak commented on module/vhost_mhu/src/mod_vhost_mhu.c:
```diff
@@ -0,0 +1,463 @@
+        return 0;
+
+    g_autoptr(GSocketAddress) addr = g_unix_socket_address_new(socket_path);
+    g_autoptr(GSocket) bind_socket = g_socket_new(G_SOCKET_FAMILY_UNIX, G_SOCKET_TYPE_STREAM,
```
seems more than 80 char

girishpathak commented on module/vhost_mhu/src/mod_vhost_mhu.c:
```diff
@@ -0,0 +1,463 @@
+    size_t chnl_idx = fwk_id_get_sub_element_idx(channel_id);
+    struct mhu_smt_channel *smt_channel = &device_ctx->smt_channel_table[chnl_idx];
+
+    vscmi_host_user_notification(&device_ctx->dev.parent, smt_channel->elem, chnl_idx, size);
```
seems more than 80 char

girishpathak commented on module/vhost_mhu/src/mod_vhost_mhu.c:
```diff
@@ -0,0 +1,463 @@
+    elt_idx = fwk_id_get_element_idx(id);
+    device_ctx = &mhu_ctx.device_ctx_table[elt_idx];
+
+    device_ctx->thread = g_thread_new(device_ctx->config->socket_path, server_wait_vm_thread, device_ctx);
```
seems more than 80 char

girishpathak commented on product/vhost-user/fw/config_ppu_v0.c:
```diff
@@ -0,0 +1,160 @@
+
+
+static const struct fwk_element *host_ppu_v0_get_element_table
+                                                           (fwk_id_t module_id)
```
alignment seems off

girishpathak reviewed (changes requested):
@vingu-linaro 
I have tried to review much of the changes, However the following commits I will review probably later this week or next week

```
750ec5758 to be sorted out
5328729a3 to be sorted out
966ec3d02 product/zephyr: create virtio scmi backend on zephyr
b6b785111 module/zephyr: create a virtio vring mailbox
937650190 arch/zephyr: add zephyr rtos architecture
782c6df9f makefile: add zephyr compilation support
ba20740a1 fwk: support context in mono threading
```
Please note below common review points

1.  Empty commit messages are not allowed, so please add at least one line in the commit message
2. Change copyright year to 2022 in all modified or new files
3. if/for/switch that requires compound statement must use {} because of Misra-c rules
4. Line more than 80 char will be rejected by CI
5. Remove commented code or else provide justification in a comment
6. If makefiles are provided make sure corresponding cmake files are also provided

Following seems work in progress commit
```
750ec5758 to be sorted out
5328729a3 to be sorted out
```
Can you update these commits with a correct commit message header?

girishpathak reviewed (changes requested):
I have tried to review much of the changes, However following commit I will review probably later this week or next week

```
750ec5758 to be sorted out
5328729a3 to be sorted out
966ec3d02 product/zephyr: create virtio scmi backend on zephyr
b6b785111 module/zephyr: create a virtio vring mailbox
937650190 arch/zephyr: add zephyr rtos architecture
782c6df9f makefile: add zephyr compilation support
ba20740a1 fwk: support context in mono threading
```
Please note below common review points

1.  Empty commit messages are not allowed, so please add at least one line in the commit message
2. Change copyright year to 2022 in all modified or new files
3. if/for/switch that requires compound statement must use {} because of Misra-c rules
4. Line more than 80 char will be rejected by CI
5. Remove commented code or else provide justification in a comment
6. If makefiles are provided make sure corresponding cmake files are also provided

Following seems work in progress commit
```
750ec5758 to be sorted out
5328729a3 to be sorted out
```
Can you update these commits with a correct commit message header

tarek-arm commented:
In addition to @girishpathak requests please consider:

1. Rebase your changes on top of the latest change.
2. The arch directory contains a lot of redundant code. Please re-use the host arch instead of creating an additional arch.
3. Split the PR into 3 requests. ( vhost - zephyr - optee)
4. For Zephyr implementation SCMI server should run as an application. There is no need to make calls into Zephyr interrupt and OsKernel functions.

vingu-linaro commented:
> I have tried to review much of the changes, However following commit I will review probably later this week or next week

Ok. Thanks

> 
> ```
> 750ec5758 to be sorted out
> 5328729a3 to be sorted out
> 966ec3d02 product/zephyr: create virtio scmi backend on zephyr
> b6b785111 module/zephyr: create a virtio vring mailbox
> 937650190 arch/zephyr: add zephyr rtos architecture
> 782c6df9f makefile: add zephyr compilation support
> ba20740a1 fwk: support context in mono threading
> ```
> 
> Please note below common review points
> 
> 1. Empty commit messages are not allowed, so please add at least one line in the commit message
I will fix this

> 2. Change copyright year to 2022 in all modified or new files
ok

> 3. if/for/switch that requires compound statement must use {} because of Misra-c rules
I have tried to follow this rule but might have missed some. I will check and fix this

> 4. Line more than 80 char will be rejected by CI
Will check this as well

> 5. Remove commented code or else provide justification in a comment
> 6. If makefiles are provided make sure corresponding cmake files are also provided
I will add missing cmake file

> 
> Following seems work in progress commit
> 
> ```
> 750ec5758 to be sorted out
> 5328729a3 to be sorted out

For these 2 commits i was not sure of the best way to handle these changes from a scp firmware PoV. 

> ```
> 
> Can you update these commits with a correct commit message header

ok

vingu-linaro commented:
> In addition to @girishpathak requests please consider:
> 
> 1. Rebase your changes on top of the latest change.

I did a rebase but didn't want to push this version because I didn't run enough tests.

> 2. The arch directory contains a lot of redundant code. Please re-use the host arch instead of creating an additional arch.
Do you mean the arch_helper.h and arch_interrupt.h files ? Should I create a  arch/none/include directory where i should put those generic header files ?

> 3. Split the PR into 3 requests. ( vhost - zephyr - optee)
Ok. I will split the PR. 

> 4. For Zephyr implementation SCMI server should run as an application. There is no need to make calls into Zephyr interrupt and OsKernel functions.

I will remove OsKernel functions. These were needed in 1st version but are not necessary anymore

For zephyr interrupt, some put/get event action needs to be uninterruptible and can happen under interrupt context. That's why I implemented them

tarek-arm commented:
> For zephyr interrupt, some put/get event action needs to be uninterruptible and can happen under interrupt context. That's why I implemented them

I am expecting you will have a wrapper zephyr application that will handle interrupts. This wrapper application can enable and disable the interrupts as it pleases. The application can trigger the event inside SCP by making a call into the fake MHU interrupt. 
This will make it more portable and platform-independent. I think you should reconsider calling it zephyr and work on the basis of creating a generic architecture that will work with any OS or bare metal. The SCP server becomes just a library that can be integrated under any environment regardless. I don't see the value from tying this into Zephyr. A generic arch will allow the user to pick whichever environment to run SCP under.

tarek-arm commented:
Again, you should look into the **_host_** arch and try to make it generic enough to meet all the requirements.

vingu-linaro commented:
> > For zephyr interrupt, some put/get event action needs to be uninterruptible and can happen under interrupt context. That's why I implemented them
> 
> I am expecting you will have a wrapper zephyr application that will handle interrupts. This wrapper application can enable and disable the interrupts as it pleases. The application can trigger the event inside SCP by making a call into the fake MHU interrupt. This will make it more portable and platform-independent. I think you should reconsider calling it zephyr and work on the basis of creating a generic architecture that will work with any OS or bare metal. The SCP server becomes just a library that can be integrated under any environment regardless. I don't see the value from tying this into Zephyr. A generic arch will allow the user to pick whichever environment to run SCP under.

I'm not sure to understand what you mean by "a wrapper zephyr application that will handle interrupts".  The interrupts are handled by zephyr which call a SCP/vring module callback but some actions on the event list must be atomic. That being said i have noticed that some BUILD_HAS_MULTITHREADING were still remaining in the code which means that I might have a hole in the atomicity of the list

vingu-linaro commented:
I have pushed a new version in which I have covered previous comments.
I have simplified the arch/none/host-optee-zephyr but they still have differences that make not possible to easily merge them in 1 single arch. I'm going to schedule a call to discuss this point.
I haven't split this PR in 3 PRs to ease the comparison between the differents architecture but will do next version

tarek-arm commented:
Jenkins: OK to test

tarek-arm commented:
Can you please run the scripts under ./tools/ to check coding style errors?
Also, can you please resolve the merge conflicts?

vingu-linaro commented:
> Can you please run the scripts under ./tools/ to check coding style errors? Also, can you please resolve the merge conflicts?

I will rebase run the scrips and rebase on latest changes

tarek-arm reviewed (changes requested):
It is not clear what is the added value from this new struct?
Just use the fwk_arch_init_driver if you have to disable interrupts or can you please add more comments to explain why this is needed

vingu-linaro commented on module/scmi_power_domain/src/mod_scmi_power_domain.c:
```diff
@@ -8,6 +8,7 @@
  *     SCMI power domain management protocol support.
  */
 
+#include <internal/scmi.h>
```
I'm not able to raise the warning anymore on latest rebase so i'm going to drop it

vingu-linaro commented:
> It is not clear what is the added value from this new struct? Just use the fwk_arch_init_driver if you have to disable interrupts or can you please add more comments to explain why this is needed

Your comment was not linked to a patch or code so I'm not sure which struct you are speaking about but I assume that your comment is about fwk_arch_atomic_driver. Is it correct ?

When scp runs with an OS, it doesn't have to manage interrupt as you mentioned in a previous review. VWe only need to ensure atomic operation when manipulating critical resources like linked list. which means that we needs 3 functions: lock/unlock and whether we run under interrupt context

By default this encapsulates the interrupt driver based mechanism in those 3 functions but an arch can provide its own version like for zephyr.

vingu-linaro commented:
I have rebased on master and removed all the warnings/errors raised by the scripts in tools

tarek-arm commented:
Currently, there is `struct fwk_arch_interrupt_driver ` which provides `int (*global_disable)(void);` This function is used for disabling the interrupts. This PR adds a new `struct fwk_arch_atomic_driver ` which also provides `unsigned int (*interrupt_lock)(void);` used for disabling the interrupts. 
I would like to understand the difference between the 2? And do we just need to rename the original `global_disable` to `global_lock` or do we need both?

tarek-arm commented on arch/none/optee/src/arch_main.c:
```diff
@@ -24,6 +24,21 @@
 #include <mod_optee_mhu.h>
+#include <internal/fwk_context.h>
+
+extern short int thread_get_id(void);
```
We have just removed multi-threading and we don't want to add code for a different implementation of multit-threading.

vingu-linaro replied:
In optee case, we have several channels which are fully independent as they don't apply on the same resources and we want to run them in parallel so we have using different event list. thread_get_id() is the optee function to know which channel is running and use the associated linked list. But there is no cross list interaction

tarek-arm commented on framework/include/fwk_core.h:
```diff
@@ -142,6 +142,22 @@ int fwk_is_delayed_response_list_empty(fwk_id_t id, bool *is_empty);
+ *
+ * \details When there are several transport channels that operate on
+ * unrelated resources, we can used dedicated execution context (event list ...)
+ * in order to parallelize the requests.
```
This looks to me like we are introducing multi-threading again!
You can have an external application that is doing this parallel resource management. The firmware can be treated as one resource and protect this single resource by means of locking externally.

vingu-linaro replied:
The goal is to prevent a sequential access and associated delay for independent channel and resources. The context is only the linked list to use with a channel and doesn't touch the other channels or lists.

vingu-linaro replied:
This is probably the most controversial part that's why I put is at the end. The goal is to be able to run independent event lists on channel which access to private resources only

vingu-linaro replied:
We have only 1 Trusted application which is the entry point for linux or u-boot to invoke scmi request. Also the notification channel still need to stay connected with request one to send back notification

vingu-linaro commented:
> Currently, there is `struct fwk_arch_interrupt_driver ` which provides `int (*global_disable)(void);` This function is used for disabling the interrupts. This PR adds a new `struct fwk_arch_atomic_driver ` which also provides `unsigned int (*interrupt_lock)(void);` used for disabling the interrupts. I would like to understand the difference between the 2? And do we just need to rename the original `global_disable` to `global_lock` or do we need both?

My understand is that struct fwk_arch_interrupt_driver is about managing at the interrupt controller level (i.e. gic). In our case, we only want to ensure some atomic operations with support of reentrancy.
As an example global_disable/enable don't care about the current/prev state and I didn't to modify this interface which targets lower level than what we need.

tarek-arm commented:
> We have only 1 Trusted application which is the entry point for linux or u-boot to invoke scmi request. Also the notification channel still need to stay connected with request one to send back notification

So are you implementing multiple event queues and one notification queue? Can you please provide some documentation explaining this new architecture?

tarek-arm commented on module/dvfs/src/mod_dvfs.c:
```diff
@@ -323,6 +329,7 @@ static int dvfs_handle_pending_request(struct mod_dvfs_domain_ctx *ctx)
     }
 
     if (ctx->config->retry_ms > 0) {
```
Using the timer is conditional. If you don't need it make sure to set retry_ms=0 in the configuration.

tarek-arm reviewed (changes requested):
I think we can safely drop this patch. Just use the configuration option to set retry_ms = 0.
The module will not use the timer. You still need to include the timer module to have access to the header file but this is the same in other cases in other modules.

tarek-arm commented on module/clock/src/mod_clock.c:
```diff
@@ -716,9 +732,10 @@ static int clock_process_event(const struct fwk_event *event,
         return clock_management_process_rate(event);
 #endif
 
+#ifdef BUILD_HAS_NOTIFICATION
```
This is also not related to notifications and should be removed

vingu-linaro replied:
I rerun some test without notification and this one is not needed anymore. This patch is old and some changes happen in clock module since

tarek-arm reviewed (changes requested):
If it helps these are the changes you need to do
`diff --git a/module/clock/src/mod_clock.c b/module/clock/src/mod_clock.c
index 7e7fe1a92..57554fe8e 100644
--- a/module/clock/src/mod_clock.c
+++ b/module/clock/src/mod_clock.c
@@ -29,7 +29,6 @@ static struct clock_ctx mod_clock_ctx;
 /*
  * Utility functions
  */
-#ifdef BUILD_HAS_NOTIFICATION
 static int process_response_event(const struct fwk_event *event)
 {
     int status;
@@ -55,7 +54,6 @@ static int process_response_event(const struct fwk_event *event)
 
     return fwk_put_event(&resp_event);
 }
-#endif /* BUILD_HAS_NOTIFICATION */
 
 static int process_request_event(const struct fwk_event *event,
                                  struct fwk_event *resp_event)
@@ -106,7 +104,6 @@ void clock_get_ctx(fwk_id_t clock_id, struct clock_dev_ctx **ctx)
     *ctx = &mod_clock_ctx.dev_ctx_table[fwk_id_get_element_idx(clock_id)];
 }
 
-#ifdef BUILD_HAS_NOTIFICATION
 /*
  * Driver response API.
  */
@@ -145,7 +142,6 @@ void clock_request_complete(
 static struct mod_clock_driver_response_api clock_driver_response_api = {
     .request_complete = clock_request_complete,
 };
-#endif
 
 /*
  * Module API functions
@@ -488,7 +484,6 @@ static int clock_process_bind_request(fwk_id_t source_id, fwk_id_t target_id,
 
         return FWK_SUCCESS;
 
-#if defined(BUILD_HAS_NOTIFICATION)
     case MOD_CLOCK_API_TYPE_DRIVER_RESPONSE:
         if (!fwk_id_is_type(target_id, FWK_ID_TYPE_ELEMENT)) {
             return FWK_E_PARAM;
@@ -503,7 +498,6 @@ static int clock_process_bind_request(fwk_id_t source_id, fwk_id_t target_id,
         }
 
         return FWK_SUCCESS;
-#endif
 
     default:
         return FWK_E_ACCESS;
@@ -732,10 +726,9 @@ static int clock_process_event(const struct fwk_event *event,
         return clock_management_process_rate(event);
 #endif
 
-#ifdef BUILD_HAS_NOTIFICATION
     case (unsigned int)CLOCK_EVENT_IDX_RESPONSE:
         return process_response_event(event);
-#endif
+
     default:
         return FWK_E_PANIC;
     }
`

tarek-arm commented on tools/check_copyright.py:
```diff
@@ -62,9 +62,9 @@
 COMPANY_FULL_NAME_PATTERN = \
-    '(Arm Limited and Contributors|Renesas Electronics Corporation)'
+    '(Arm Limited and Contributors|Renesas Electronics Corporation|Linaro Limited and Contributors)'
 
```
We would like to limit the line to 80 characters as mentioned in the code style doc

tarek-arm reviewed (changes requested):
You can say in the commit message "Add Linaro as a contributor to the SCP project."

nicola-mazzucato-arm commented:
Close for now, thanks!
